src/main/user-impl/java/com/mysql/cj/jdbc/ha/plugins/failover/ClusterAwareMetricsContainer.java

```diff
@@ -0,0 +1,218 @@
+ * this list of conditions and the following disclaimer in the documentation and/or
+ * other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
+ * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT
+ * SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ */
+
+package com.mysql.cj.jdbc.ha.plugins.failover;
+
+import com.mysql.cj.conf.PropertyKey;
+import com.mysql.cj.conf.PropertySet;
+import com.mysql.cj.jdbc.JdbcConnection;
+import com.mysql.cj.jdbc.ha.plugins.ICanCollectPerformanceMetrics;
+import com.mysql.cj.jdbc.ha.plugins.ICurrentConnectionProvider;
+import com.mysql.cj.log.Log;
+
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+public class ClusterAwareMetricsContainer implements IClusterAwareMetricsContainer {
+    // ClusterID, Metrics
+    private final static Map<String, ClusterAwareMetrics> clusterMetrics = new ConcurrentHashMap<>();
+    private final static Map<String, ClusterAwareTimeMetricsHolder> topologyMetrics = new ConcurrentHashMap<>();
+    // Instance URL, Metrics
+    private final static Map<String, ClusterAwareMetrics> instanceMetrics = new ConcurrentHashMap<>();
+
+    private ICurrentConnectionProvider currentConnectionProvider;
+    private PropertySet propertySet;
+    private String clusterId = "";
+
+    public ClusterAwareMetricsContainer(ICurrentConnectionProvider currentConnectionProvider, PropertySet propertySet) {
+        this.currentConnectionProvider = currentConnectionProvider;
+        this.propertySet = propertySet;
+    }
+
+    public void setClusterId(String clusterId) {
+        this.clusterId = clusterId;
+    }
+
+    public void registerTopologyQueryExecutionTime(String clusterId, long timeMs) {
+        topologyMetrics.computeIfAbsent(clusterId, k -> new ClusterAwareTimeMetricsHolder("Topology Query"))
+            .registerQueryExecutionTime(timeMs);
+    }
+
+    public void registerFailureDetectionTime(long timeMs) {
+        if (!isEnabled()) {
+            return;
+        }
+
+        getClusterMetrics(clusterId).registerFailureDetectionTime(timeMs);
+
+        if (isInstanceMetricsEnabled()) {
+            getInstanceMetrics(getCurrentConnUrl()).registerFailureDetectionTime(timeMs);
+        }
+    }
+
+    public void registerWriterFailoverProcedureTime(long timeMs) {
```

Comment:
This logic can be improved by extracting a method and accepting a lambda to register an event.